commit 9d1887c28c05
Author: Thomas Nguyen <tnguyen@mozilla.com>
Date:   Tue Aug 22 14:51:44 2017 +0800

    Bug 1363038 - Remove synchronous dispatch when worker is racing against shutdown r=hchang
    
    MozReview-Commit-ID: JOkOukz1fov
    
    --HG--
    extra : rebase_source : 9e4204cebcd4615078019a51818250ff8125efe1
---
 .../url-classifier/nsUrlClassifierDBService.cpp     | 21 ++++++++-------------
 .../url-classifier/nsUrlClassifierDBService.h       |  3 +++
 2 files changed, 11 insertions(+), 13 deletions(-)

diff --git toolkit/components/url-classifier/nsUrlClassifierDBService.cpp toolkit/components/url-classifier/nsUrlClassifierDBService.cpp
index 3f7e3c8daaac..074bfc583b93 100644
--- toolkit/components/url-classifier/nsUrlClassifierDBService.cpp
+++ toolkit/components/url-classifier/nsUrlClassifierDBService.cpp
@@ -1616,12 +1616,6 @@ nsUrlClassifierDBService::Init()
   ReadTablesFromPrefs();
   nsresult rv;
 
-  {
-    // Force PSM loading on main thread
-    nsCOMPtr<nsICryptoHash> dummy = do_CreateInstance(NS_CRYPTO_HASH_CONTRACTID, &rv);
-    NS_ENSURE_SUCCESS(rv, rv);
-  }
-
   {
     // Force nsIUrlClassifierUtils loading on main thread.
     nsCOMPtr<nsIUrlClassifierUtils> dummy =
@@ -2278,13 +2272,14 @@ nsUrlClassifierDBService::Shutdown()
   //    is to avoid racing for Classifier::mUpdateThread
   //    between main thread and the worker thread. (Both threads
   //    would access Classifier::mUpdateThread.)
-  using Worker = nsUrlClassifierDBServiceWorker;
-  RefPtr<nsIRunnable> r = NewRunnableMethod(
-    "nsUrlClassifierDBServiceWorker::FlushAndDisableAsyncUpdate",
-    mWorker,
-    &Worker::FlushAndDisableAsyncUpdate);
-  SyncRunnable::DispatchToThread(gDbBackgroundThread, r);
-
+  if (mWorker->IsDBOpened()) {
+    using Worker = nsUrlClassifierDBServiceWorker;
+    RefPtr<nsIRunnable> r = NewRunnableMethod(
+      "nsUrlClassifierDBServiceWorker::FlushAndDisableAsyncUpdate",
+      mWorker,
+      &Worker::FlushAndDisableAsyncUpdate);
+    SyncRunnable::DispatchToThread(gDbBackgroundThread, r);
+  }
   // At this point the update thread has been shut down and
   // the worker thread should only have at most one event,
   // which is the callback event.
diff --git toolkit/components/url-classifier/nsUrlClassifierDBService.h toolkit/components/url-classifier/nsUrlClassifierDBService.h
index 13987b0ed2ee..8d284f5b0842 100644
--- toolkit/components/url-classifier/nsUrlClassifierDBService.h
+++ toolkit/components/url-classifier/nsUrlClassifierDBService.h
@@ -227,6 +227,9 @@ public:
   // will be nulled out in NotifyUpdateObserver.
   bool IsBusyUpdating() const { return !!mUpdateObserver; }
 
+  // Check the DB ready state of the worker thread
+  bool IsDBOpened() const { return !!mClassifier; }
+
   // Delegate Classifier to disable async update. If there is an
   // ongoing update on the update thread, we will be blocked until
   // the background update is done and callback is fired.
